idkit/src/services/wld-bridge.ts

```diff
@@ -0,0 +1,175 @@
+import { create } from 'zustand'
+import type { ISuccessResult } from '..'
+import { useEffect, useRef } from 'react'
+import { buffer_decode } from '@/lib/utils'
+import type { IDKitConfig } from '@/types/config'
+import { VerificationState } from '@/types/bridge'
+import type { AppErrorCodes } from '@/types/bridge'
+import { decryptResponse, encryptRequest, exportKey, generateKey } from '@/lib/crypto'
+import { encodeAction, generateExternalNullifier, generateSignal } from '@/lib/hashing'
+
+const DEFAULT_BRIDGE_URL = 'https://bridge.worldcoin.org'
+
+type WorldBridgeStore = {
+	bridgeUrl: string
+	iv: Uint8Array | null
+	key: CryptoKey | null
+	requestId: string | null
+	connectorURI: string | null
+	result: ISuccessResult | null
+	errorCode: AppErrorCodes | null
+	verificationState: VerificationState
+
+	createClient: (
+		app_id: IDKitConfig['app_id'],
+		action: IDKitConfig['action'],
+		signal?: IDKitConfig['signal'],
+		bridgeUrl?: IDKitConfig['bridgeUrl'],
+		credential_types?: IDKitConfig['credential_types'],
+		action_description?: IDKitConfig['action_description']
+	) => Promise<void>
+
+	pollForUpdates: () => Promise<void>
+
+	reset: () => void
+}
+
+const useWorldBridgeStore = create<WorldBridgeStore>((set, get) => ({
+	iv: null,
+	key: null,
+	result: null,
+	errorCode: null,
+	requestId: null,
+	connectorURI: null,
+	bridgeUrl: DEFAULT_BRIDGE_URL,
+	verificationState: VerificationState.PreparingClient,
+
+	createClient: async (
+		app_id: IDKitConfig['app_id'],
+		action: IDKitConfig['action'],
+		signal?: IDKitConfig['signal'],
+		bridgeUrl?: IDKitConfig['bridgeUrl'],
+		credential_types?: IDKitConfig['credential_types'],
+		action_description?: IDKitConfig['action_description']
+	) => {
+		const { key, iv } = await generateKey()
+		const requestId = window.crypto.randomUUID()
+
+		const res = await fetch(`${bridgeUrl ?? DEFAULT_BRIDGE_URL}/request/${requestId}`, {
+			method: 'PUT',
+			headers: { 'Content-Type': 'application/json' },
+			body: JSON.stringify(
+				await encryptRequest(
+					key,
+					iv,
+					JSON.stringify({
+						app_id,
+						credential_types,
+						action_description,
+						action: encodeAction(action),
+						external_nullifier: generateExternalNullifier(app_id, action).digest,
```

Comment:
@m1guelpf we need to pass the external_nullifier to the app for the advanced on-chain cases and cases where devs aren't using the Dev Portal to register their apps/actions. however, I'm not sure if for the advanced on-chain use cases, this is the best way to do it, will devs rely on creating a "custom action" for this case?